File: backend/app/core/permissions.py
from __future__ import annotations


PERMISSION_ACTIONS = ("view", "create", "edit", "delete")

PERMISSION_MODULES = (
    "vehicles",
    "possession",
    "drivers",
    "maintenance",
    "claims",
    "fines",
    "master_data",
    "fuel_supplies",
    "fuel_supply_orders",
    "fuel_stations",
    "payment_processes",
    "analytics",
    "data_imports",
)

POSSESSION_ROLE_CEILINGS = {
    "ADMIN": {"can_view": True, "can_create": True, "can_edit": True, "can_delete": False},
    "PRODUCAO": {"can_view": True, "can_create": True, "can_edit": True, "can_delete": False},
    "PADRAO": {"can_view": True, "can_create": False, "can_edit": False, "can_delete": False},
    "POSTO": {"can_view": False, "can_create": False, "can_edit": False, "can_delete": False},
}


def blank_permissions() -> dict[str, dict[str, bool]]:
    return {
        module: {
            "can_view": False,
            "can_create": False,
            "can_edit": False,
            "can_delete": False,
        }
        for module in PERMISSION_MODULES
    }


def default_permissions_for_role(role: str) -> dict[str, dict[str, bool]]:
    permissions = blank_permissions()

    if role == "ADMIN":
        for module in permissions:
            permissions[module] = {
                "can_view": True,
                "can_create": True,
                "can_edit": True,
                "can_delete": True,
            }
        permissions["possession"] = POSSESSION_ROLE_CEILINGS[role].copy()
        return permissions

    if role == "PRODUCAO":
        for module in (
            "vehicles",
            "possession",
            "drivers",
            "maintenance",
            "claims",
            "fines",
            "master_data",
            "fuel_supplies",
            "fuel_supply_orders",
            "payment_processes",
            "data_imports",
        ):
            permissions[module].update(can_view=True, can_create=True, can_edit=True)
        permissions["fuel_stations"]["can_view"] = True
        return permissions

    if role == "POSTO":
        permissions["fuel_supply_orders"].update(can_view=True, can_edit=True)
        return permissions

    if role == "PADRAO":
        permissions["possession"] = POSSESSION_ROLE_CEILINGS[role].copy()
        return permissions

    return permissions


def apply_role_permission_ceiling(
    role: str,
    module: str,
    flags: dict[str, bool],
) -> dict[str, bool]:
    if module != "possession":
        return flags.copy()

    ceiling = POSSESSION_ROLE_CEILINGS.get(role, blank_permissions()["possession"])
    return {key: bool(flags.get(key, False) and ceiling[key]) for key in ceiling}


def action_to_column(action: str) -> str:
    if action not in PERMISSION_ACTIONS:
        raise ValueError(f"Ação de permissão desconhecida: {action}")
    return f"can_{action}"

File: frontend/src/utils/permissions.js
export const PERMISSION_MODULES = [
  { key: 'vehicles', label: 'Veículos' },
  { key: 'possession', label: 'Posses' },
  { key: 'drivers', label: 'Condutores' },
  { key: 'maintenance', label: 'Manutenções' },
  { key: 'claims', label: 'Sinistros' },
  { key: 'fines', label: 'Multas' },
  { key: 'master_data', label: 'Cadastros' },
  { key: 'fuel_supplies', label: 'Abastecimentos' },
  { key: 'fuel_supply_orders', label: 'Ordens de abastecimento' },
  { key: 'fuel_stations', label: 'Postos' },
  { key: 'payment_processes', label: 'Processos de pagamento' },
  { key: 'analytics', label: 'Analytics' },
  { key: 'data_imports', label: 'Importação/Exportação' },
]

export const PERMISSION_ACTIONS = [
  { key: 'view', field: 'can_view', label: 'Visualizar' },
  { key: 'create', field: 'can_create', label: 'Criar' },
  { key: 'edit', field: 'can_edit', label: 'Editar' },
  { key: 'delete', field: 'can_delete', label: 'Excluir' },
]

const EMPTY_FLAGS = {
  can_view: false,
  can_create: false,
  can_edit: false,
  can_delete: false,
}

export function normalizePermissions(permissions = {}) {
  return PERMISSION_MODULES.reduce((acc, module) => {
    const flags = permissions?.[module.key] || EMPTY_FLAGS
    acc[module.key] = {
      can_view: Boolean(flags.can_view),
      can_create: Boolean(flags.can_create),
      can_edit: Boolean(flags.can_edit),
      can_delete: Boolean(flags.can_delete),
    }
    return acc
  }, {})
}

export function hasPermission(user, module, action = 'view') {
  const field = action.startsWith('can_') ? action : `can_${action}`
  return Boolean(user?.permissions?.[module]?.[field])
}
